src/awkward/operations/ak_from_buffers.py

```diff
@@ -123,11 +125,27 @@ def _impl(
     return wrap_layout(out, behavior, highlevel)
 
 
-def _from_buffer(nplike, buffer, dtype, count, byteorder):
-    if nplike.is_own_array(buffer):
+def _from_buffer(
+    nplike: NumpyLike, buffer, dtype: np.dtype, count: ShapeItem, byteorder: str
+) -> ArrayLike:
+    # Unknown-length information implies that we didn't load shape-buffers (offsets, etc)
+    # for the parent of this node. Thus, this node and its children *must* only
+    # contain placeholders
+    if count is unknown_length:
+        if not isinstance(buffer, PlaceholderArray):
+            raise AssertionError("Encountered unknown length for concrete buffer")
+        return PlaceholderArray(nplike, (unknown_length,), dtype)
+    # Known-length information implies that we should have known-length buffers here
+    # Therefore, placeholders without shape information are not permitted
+    elif isinstance(buffer, PlaceholderArray) or nplike.is_own_array(buffer):
+        # Require 1D buffers
         array = nplike.reshape(buffer.view(dtype), shape=(-1,), copy=False)
```

Comment:
It's good that this can all happen in one place (i.e. that there's a `_from_buffer` function).